lib/new_relic/agent/instrumentation/aws_sdk_lambda/instrumentation.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+require 'json'
+
+module NewRelic::Agent::Instrumentation
+  module AwsSdkLambda
+    INSTRUMENTATION_NAME = 'aws_sdk_lambda'
+    AWS_SERVICE = 'lambda'
+    CLOUD_PLATFORM = 'aws_lambda'
+    WRAPPED_RESPONSE = Struct.new(:status_code, :has_status_code?)
+
+    def invoke_with_new_relic(*args)
+      with_tracing(:invoke, *args) { yield }
+    end
+
+    def invoke_async_with_new_relic(*args)
+      with_tracing(:invoke_async, *args) { yield }
+    end
+
+    def invoke_with_response_stream_with_new_relic(*args)
+      with_tracing(:invoke_with_response_stream, *args) { yield }
+    end
+
+    private
+
+    def with_tracing(action, *args)
+      segment = generate_segment(action, *args)
+
+      # prevent additional instrumentation for things like Net::HTTP from
+      # creating any segments that may appear as redundant / confusing
+      NewRelic::Agent.disable_all_tracing do
+        begin
+          response = yield
+          process_response(response, segment)
+          response
+        rescue => e
+          NewRelic::Agent.notice_error(e)
+          raise
+        end
+      end
+    ensure
+      segment&.finish
+    end
+
+    def process_response(response, segment)
+      process_status_code(response, segment) if response.respond_to?(:status_code)
+      process_function_error(response) if response.respond_to?(:function_error)
+    end
+
+    def process_status_code(response, segment)
+      status_code = response.status_code
+      return unless status_code
+
+      segment.process_response_headers(WRAPPED_RESPONSE.new(status_code, true))
+    end
+
+    # notice error that was raised / unhandled by the function
+    def process_function_error(response)
+      function_error = response.function_error
+      return unless function_error
+
+      msg = "[#{function_error}]"
+      payload = response.payload&.string if response.respond_to?(:payload)
+      payload_hash = JSON.parse(payload) if payload
+      msg = "#{msg} #{payload_hash['errorType']} - #{payload_hash['errorMessage']}" if payload_hash
+      e = StandardError.new(msg)
+      e.set_backtrace(payload_hash['stackTrace']) if payload_hash
+
+      NewRelic::Agent.notice_error(e)
+    end
+
+    def generate_segment(action, options = {})
+      function = function_name(options)
+      region = aws_region
+      account_id = aws_account_id
+      arn = aws_arn(function, region)
+
+      segment = NewRelic::Agent::Tracer.start_external_request_segment(
+        library: INSTRUMENTATION_NAME,
+        uri: "https://lambda.#{aws_region || 'unknown-region'}.amazonaws.com",
+        procedure: action
+      )
+      segment.name = "External/Lambda/#{action}/#{function}"
+      segment.add_agent_attribute('cloud.account.id', account_id)
+      segment.add_agent_attribute('cloud.platform', CLOUD_PLATFORM)
+      segment.add_agent_attribute('cloud.region', region)
+      segment.add_agent_attribute('cloud.resource_id', arn) if arn
```

Comment:
There's a guard for `arn`, but not for `account_id` or `region`. Is it okay to have nil/empty values for the other attributes?

Reply:
Yeah, my interpretation of the spec and existing aws-sdk related instrumentation is that `arn` is the only one that gets a guard. Further spec refinement might end up clarifying that the others need guards too, but for now `arn` was the only one with a specific instruction to guard.